Adding bylaws

sffc opened the pull request.

hagbard commented on docs/bylaws.md:
```diff
@@ -0,0 +1,51 @@
+
+- Email
+- Slack
+- Google docs
```
I think docs should be a first class thing for discussions. They have unique features above any other listed "channel".
I certainly intend to use the tool that makes things easiest, so I expect to create and collaborate on many technical docs which might impact direction/design. Happy to later copy them somewhere once collaboration phase is over.

sffc replied:
I'm personally not convinced that what Google Docs brings to the table outweighs the cost of splitting technical discussions between two platforms.

The two key features of Google Docs, document layout and discussion threads, can be done on GitHub via Issues to discuss specific questions and Markdown files to add additional context.  GitHub keeps those two concerns separate, so that open questions can be tracked separately from the working draft.

GitHub issues can be linked; if we have too many technical discussions on Google Docs, we are more likely to duplicate threads.

Also, a big deficiency with Google Docs is that discussion threads are easily lost.

sffc replied:
To be clear, I intentionally proposed "rapid prototyping" as a catch-all use case for the back channels, including Google Docs.  If it's a format that's useful for brainstorming ideas, then that's fine as rapid prototyping.  However, in my proposed bylaws, the doc should be moved to a Markdown file on GitHub before it is considered as having been adopted as a draft by the subcommittee.  While in Google Doc form, it's merely the author's personal brainstorm of ideas.

Does that sounds reasonable?

Manishearth replied:
Yeah I find Google Docs to be pretty lacking for this kind of thing. For brainstorming, yeah, but aside from that I prefer discussions-first platforms.

hagbard replied:
Being able to highlight 2 words in a proposed bit of code/docs and say "why this thing specifically?" in a comment thread alongside (and then do that in 20 places in the same doc) is so much more efficient in terms of establishing context to the people in the discussion than having 20 issues or one issue with lots of intermixing of threads.

Yes, you can go into the "source" and do reviews in PRs, but it lacks WYSIWYG and is cumbersome compared to just "highlight and comment". Having got very used to it for all the internal API reviews I'm involved with (10 people adding, editing and resolving comments in real time during the review) I'd say it's 2 orders of magnitude better (efficiency wise) for collaboration than async review threads.

You need some principles for collaborating like this (Kevin Bourillon has some and I might see if I can post them here) but it's astonishingly effective at reaching consensus and working with new ideas.

sffc replied:
I think your use case is covered by the "Rapid Prototyping" exception for Google Docs.

If you make a Google Doc, people can review it for correctness, highlight one or two words with small suggestions, etc.  However, it's natural that your doc will have several larger open questions.  In that case, you should explicitly list those open questions clearly in your doc.

That's the "Rapid Prototyping" phase.  After that's complete, the next step is to copy the doc to a Markdown file in a PR on GitHub, open issues for your open questions, and the PR gets checked in (pending final approval from the SC).  We can then track those open questions along with everyone else's open questions until they are resolved, which is done as subsequent PRs against the Markdown file.

sffc replied:
I have documented this process in #21.

echeran commented on docs/bylaws.md:
```diff
@@ -0,0 +1,51 @@
+
+### Technical Decisions
+
+OmnICU-TC maintains a collection of explainers (design documents) that are used as a basis for technical decisions.  All explainers that have been approved by OmnICU-SC live as Markdown files on GitHub.  *Designs are not official unless they are in an explainer checked-in to GitHub.*
```
Could there be a separate document from these bylaws that gives tips / howto for contributing?  And maybe it would start off only containing a link to a good tool for making the SVG diagrams for explainers, and how people view such diagrams in an explainer PR.

sffc replied:
That's a good idea.  I'll do it in a follow-up PR.

sffc replied:
Done: #21

sffc commented:
We heard consensus on these bylaws in today's OmnICU meeting.

There were some concerns over the categorization of Google Docs as "back channels", but my clarification of the Rapid Prototyping exception addressed those concerns, and further, @Manishearth and @hsivonen advocated in favor of the proposal because:

1. If you leave your company, the Google Docs you made could go away.
2. Permissions sharing on docs are complicated; even if you think you have the doc shared publicly, you might not, especially if you are using it through a GSuite account where your admin has restrictive sharing policies in place that overrule your own sharing policies (as is the case for Google employees).
3. Google Docs get lost easily; a central listing would be required if we use them as public record.
4. It is easier to track the evolution of designs over time when you can review commits and PRs.
5. Tracking conversations is easier when everything is in one place on GitHub.

In the end, there may be ways to work around these issues, but I would like us to give the GitHub-first approach a try, and we can always modify our processes later if there is sufficient need.

sffc commented:
I have made an additional document with process tips in #21.  I think we can merge in the Bylaws now.
